### PR TITLE
feat: Add `release.sh` script and detect breaking changes

### PR DESCRIPTION
**Pull request by mafredri**
**NOTE**: This is not the final form of how releases will be made, see below.

This commit introduces three new scripts:

- `release.sh` To be run by a user on their local machine to preview and
  create a new release (tag + push)
- `check_commit_metadata.sh` For e.g. detecting breaking changes
- `genereate_release_notes.sh` To display the generated release notes,
  used for previews and in `publish_release.sh`

The `release.sh` script can be run without arguments, and it will
automatically determine if we're to do a patch or minor release. A minor
release can be forced via `--minor` flag.

Breaking changes can be annotated either via commit/merge title prefix
(`feat!:`, `feat(api)!:`), or by adding the `release/breaking` label to
the PR that was merged (on GitHub).

Related #5233

This work will be followed up by changes to move the tag creation from
`release.sh` to CI workflows, the `release.sh` will then become a
lightweight wrapper to run the workflow.

Documented here: #5381, #5382

---

Example output:

```console
❯ ./scripts/release.sh
Fetching main and tags from origin...
Checking commit metadata for changes since v0.13.1...
(faked) Breaking change detected, incrementing minor version...
Old version: v0.13.1
New version: v0.14.0

Preview release notes? (y/n)
## Changelog:

### BREAKING CHANGES

- 81c39487 fix: Close tty first in `ptytest` cleanup (#5252)

### Features

- 061635c3 feat: Allow multiple OIDC domains (#5210)
- 9e4d213c feat: add lazygit, remove kubic, use dotfiles in dogfood image (#5271)
- e17fd0bb feat: Add GET previous template version endpoint (#5230)
- ce76d9d5 feat: Add diff and Dockerfile support for template version page (#5339)

### Bug fixes

- 92c217bd fix: add index on `workspace_agents.auth_token` (#5244)
- 8469dbc0 fix: add index to `provisioner_jobs.started_at` (#5245)
- fa641554 fix: Improve agent connection tracking when agent is closed (#5253)
- 137a48c2 fix: workspaceapps: overloaded test server responds with 502s (#5255)
- a02617b6 fix: remove unnecessary `WHERE`s from `AcquireProvisionerJob` (#5257)
- cec667d3 fix: prettier misses docs directory (#5285)
- e04877a6 fix: race conditions in replicasync (#5289)
- 02dcd0e2 fix: Fix resource avatar when icon is empty string (#5291)
- d3200382 fix: agent panics on closed network (#5295)
- 84872d97 fix: loadtest/reconnectingpty tweak timeout (#5300)
- b2dc60c0 fix: markdown-link-check base-branch should not be set on main branch (#5311)
- c77c1b4b fix: Retry if there is no git auth user yet (#5316)
- 6651c163 fix: avoid terraform state concurrent access, remove global mutex (#5273)
- 161465db fix: do not canonicalize Sec-WebSocket-* headers in apps (#5334)
- 2f3ff6ce fix: improve pty and ptytest (#5327)
- 3cea5f96 fix: wait for creating template versions (#5335)
- f7467cac fix: Improve ptytest closure on expect match timeout (#5337)
- ee605b34 fix: Don't show progress bar for new templates (#5298)
- 8ea09235 fix: UX issues in template settings form's default auto-stop field (#5330)
- 534bff2f fix: coderd/prometheusmetrics wait for all metrics in require.Eventually (#5338)
- c063ac24 fix: use doWithRetries when making HTTP calls (#5344)
- 687261c8 fix: Close coordinator in wsconncache test (#5348)
- f68a6569 fix: winget package releases (#5352)
- 3c9dab34 fix: Fix CSP for monaco editor (#5358)
- 92c5e97f fix: Fix CSP style directive for Monaco editor (#5360)
- 05130db5 fix: Improve closing of services in agent tests (#5355)
- 935d2eb5 fix: fmt should check for unstaged files (#5362)
- cd04330c fix: replace fireEvent with userEvent (#5361)

### Other changes

- 65407462 Add audit links/kira pilot (#5156)
- df389d42 Add build number to workspace_build audit logs (#5267)
- 1cfe5de1 Add Service Banners (#5272)
- 02bb052d Fix template Avatar (#5294)
- ee74df3d Fix scope of dbTTL (#5197)
- ee4f0fc5 chore: enable debug logging for gotestsum (#5248)
- 133b2de1 chore: improve markdown-link-check workflow (#5303)
- 825480ae chore: bump crate-ci/typos from 1.12.12 to 1.13.3 (#5304)
- 4a0ca481 chore: bump @typescript-eslint/parser from 5.38.1 to 5.45.1 in /sit (#5305)
- 03328d4f chore: bump eslint from 8.24.0 to 8.29.0 in /site (#5306)
- a973c35a chore: collect gotestsum TestEvents as workflow artifacts (#5336)
- 1ac1af7d chore: bump emoji-mart from 5.2.1 to 5.3.3 in /site (#5226)
- 3d95c925 chore: Add DatoCMS token to dogfood template (#5254)
- 7eb3ab04 chore: bump decode-uri-component from 0.2.0 to 0.2.2 in /site (#5251)
- 971e3678 chore: improve logging in provisionerd_test (#5353)
- 91973e1e cli: remove redundant client creation requests (#5264)
- a3083f77 docs: fix community templates link on site. (#5278)
- b4603582 docs: add templates link to README.md (#5201)
- 59af8349 docs: add offical kubernetes provider runtime_class_name (#5157)
- 85945af5 fixed bug; wrote tests (#5329)
- fd545128 helm: add certs secret mount (#4641)
- 9cfdbec2 refactor: Remove login banner (#5239)
- 85a6d14f skip: reconnectingpty tests (#5322)

https://github.com/coder/coder/compare/v0.13.1...v0.14.0

## Container image:

- `docker pull ghcr.io/coder/coder:v0.14.0`

Create release? (y/n)
Tagging commit 971e3678 as v0.14.0...
git tag -a v0.14.0 -m v0.14.0 971e3678
Pushing tag to origin...
git push -u origin v0.14.0
```

### CODE REVIEW

File: scripts/release.sh [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	echo -e "$release_notes\n"
+fi
+
+read -p "Create release? (y/n) " -n 1 -r create
+echo
+if [[ $create =~ ^[Yy]$ ]]; then
+	log "Tagging commit $ref as $new_version..."
+	echo git tag -a "$new_version" -m "$new_version" "$ref"
+	log "Pushing tag to origin..."
+	echo git push -u origin "$new_version"
```
Comment (mafredri) [inline scripts/release.sh:111]: Review: Echo for the paranoid... 😄

**Comment by bpmct** [comment]
Looks great!! What do you think about failing the release workflow if the user did a manual `git tag && git push`? @kylecarbs and I are the only releasers right now, so it might not be a big deal but it might not hurt to systematically avoid manual releases now that we have the script :)

**Comment by mafredri** [comment]
> Looks great!! What do you think about failing the release workflow if the user did a manual `git tag && git push`? @kylecarbs and I are the only releasers right now, so it might not be a big deal but it might not hurt to systematically avoid manual releases now that we have the script :)

I'm definitely supportive of that idea. Do you think we should have it in this PR already or is it OK to leave to a follow up one? I'm going to do some workflow changes and a bit more refactoring of `publish_release.sh` in a follow up PR and it would be a good fit there.

One easy way to throw it into this PR would be to change the tag message `-m 'Release vX.X.X (via release.sh)'` or something similar, and then skip it if it doesn't contain the right message format.

**Comment by bpmct** [comment]
> Do you think we should have it in this PR already or is it OK to leave to a follow up one?

A follow-up one (and issue) makes sense! I wouldn't consider it a blocker at all :)

File: scripts/check_commit_metadata.sh [hunk 2]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env bash
+
+# Usage: source ./check_commit_tags.sh <revision range>
+# Usage: ./check_commit_tags.sh <revision range>
+#
+# Example: ./check_commit_tags.sh v0.13.1..971e3678
+#
+# When sourced, this script will populate the COMMIT_METADATA_* variables
```
Comment (mtojek) [inline scripts/check_commit_metadata.sh:8]: _Applicable only to squash & merge_ - Do you think that we may need another script executed in CI to check if the title of PR is consistent?

File: scripts/generate_release_notes.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	fi
+)"
+
+image_tag="$(execrelative ./image_tag.sh --version "$new_version")"
+
+echo -e "## Changelog:
+
+$changelog
+
+https://github.com/coder/coder/compare/${old_version}...${new_version}
```
Comment (mtojek) [inline scripts/generate_release_notes.sh:116]: nit: md link?

**Review by mtojek — COMMENTED** [review]
My biggest concern is lack of tests to validate all cases of generating the changelog, but I don't insist on enabling any test runner like `bats`. I guess that it might be hard to integrate with `gh api`.

File: scripts/generate_release_notes.sh (hunk 3 above)
Comment (mafredri) [reply scripts/generate_release_notes.sh:116]: This will be automatically formatted as `v0.13.1...v0.13.2` by GitHub, I guess we could make it explicit if we want?

File: scripts/check_commit_metadata.sh (hunk 2 above)
Comment (mafredri) [reply scripts/check_commit_metadata.sh:8]: Hmm, what do you mean by _Applicable only to squash & merge_? I guess we could make it clearer this script is only intended for the commit log on the `main` branch.

Yes, we'll add an action/lint for PR title 👍🏻

File: scripts/generate_release_notes.sh (hunk 3 above)
Comment (mtojek) [reply scripts/generate_release_notes.sh:116]: I was thinking about using the [link](https://github.com/coder/coder/compare/v0.13.1...v0.13.2) instead. I'm fine with formatting. 

BTW it's just a nit-pick, so feel free to ignore.

File: scripts/check_commit_metadata.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		# Get the labels for the PR associated with this commit.
+		mapfile -t labels < <(gh api -H "Accept: application/vnd.github+json" "/repos/coder/coder/commits/${commit_sha}/pulls" -q '.[].labels[].name')
+
+		if [[ " ${labels[*]} " = *" ${breaking_label} "* ]]; then
+			COMMIT_METADATA_CATEGORY[$commit_sha]=$breaking_category
+			COMMIT_METADATA_BREAKING=1
+			continue
+		fi
+
+		if [[ $commit_prefix =~ $prefix_pattern ]]; then
```
[the remark this answers is not in the document]
Comment (mafredri) [reply scripts/check_commit_metadata.sh:71]: I won't get behind Python, but I'm indifferent to mage, so it's not me you have to convince. 😉

File: scripts/check_commit_metadata.sh (hunk 2 above)
Comment (mtojek) [reply scripts/check_commit_metadata.sh:8]: > Yes, we'll add an action/lint for PR title 👍🏻 (follow up PR)

👍 

> Applicable only to squash & merge?

I thought that `Create a merge commit` option is also enabled for this repo, but it seems to be off. If it's on, it might be more challenging to validate also commits than only the PR title.

**Comment by mafredri** [comment]
> My biggest concern is lack of tests to validate all cases of generating the changelog, but I don't insist on enabling any test runner like `bats`. I guess that it might be hard to integrate with `gh api`.

It's not necessarily hard per-se, we can just `gh() { if ((testing)); then gh_fake "$@"; else command gh "$@"; fi }`, but I don't think it will bring a whole lot of value for a few reasons:

- The changelog is generated the same way locally and in CI, and it can be previewed by the releaser (I could change to always print the changelog instead of optionally)
- The changelog is editable after the face (if something went wrong)
- If we encounter an error, release will abort (`set -e`)

I guess the biggest risk is in potentially not detecting a breaking change and only doing patch release if something goes wrong or a future change in the script breaks it.

File: scripts/check_commit_metadata.sh (hunk 2 above)
Comment (mafredri) [reply scripts/check_commit_metadata.sh:8]: Ah right, yeah we'll only validate PR titles, I made the assumption when writing that we'll only squash merge.

Obviously owners/admins could bypass this by pushing straight to `main`. 🙄

File: scripts/check_commit_metadata.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (mafredri) [reply scripts/check_commit_metadata.sh:1]: I guess we have enough release related scripts to put them there, I'd still like to keep `/scripts/release.sh` (because used by humans) but the rest can go in `/scripts/release/*.sh`.

**Review by bpmct — APPROVED** [review]
I just tested it and all works as expected! 

(I manually copied the changelog since I understand the script will not do the new changelog  format until we merge the new workflow)

File: scripts/release/check_commit_metadata.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		# First, check the title for breaking changes. This avoids doing a
+		# GH API request if there's a match.
+		if [[ $commit_prefix =~ $breaking_title ]]; then
+			COMMIT_METADATA_CATEGORY[$commit_sha]=$breaking_category
+			COMMIT_METADATA_BREAKING=1
+			continue
+		fi
+
+		# Get the labels for the PR associated with this commit.
+		mapfile -t labels < <(gh api -H "Accept: application/vnd.github+json" "/repos/coder/coder/commits/${commit_sha}/pulls" -q '.[].labels[].name')
```
Comment (deansheather) [inline scripts/release/check_commit_metadata.sh:63]: The rate limit for calls from github actions is 1k per hour per repository. I'd say an upper bound for the amount of PRs in a release would be 100, but combined with other things in the repo we might be pushing on 1k per hour during busy periods. Can this instead use a list endpoints?

The commits we release are a contiguous block on main so if you list from first ref to end ref then you will get them all in much less requests.

**Review by deansheather — APPROVED** [review]
These scripts seem good but they are hard to follow even for me. We're starting to reach a point where it'd be better to write these more complex scripts in a different language like python or go

File: scripts/lib.sh [hunk 7]
```diff
@@ -123,6 +123,9 @@ log() {
 # error prints an error message and returns an error exit code.
 error() {
 	log "ERROR: $*"
+	if issourced; then
+		return 1
+	fi
```
[the remark this answers is not in the document]
Comment (mafredri) [reply scripts/lib.sh:128]: Actually, I'll just remove it. It's not really necessary if we always want to exit on `error`. I had the scripts set up in a different way before.

File: scripts/release/check_commit_metadata.sh (hunk 6 above)
Comment (mafredri) [reply scripts/release/check_commit_metadata.sh:63]: I did consider this but it's not clear how we'd trigger it. Now I don't remember but do we run dry-run for every merged commit? I considered adding a `--skip-label-check` to this script used in dry-run scenarios, but omitted it for simplicity and because it would generate "false" release notes.

> The commits we release are a contiguous block on main so if you list from first ref to end ref then you will get them all in much less requests.

The problem isn't listing the commits, but in that endpoint they won't include the metadata we're looking for (PR labels). I tried to find a way to get all of this information in one go, but each one has caveats.. 😅

We could use [`/repos/coder/coder/pulls`](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests), but even with filtering `?state=closed&base=main&per_page=100` (maybe `sort=updated`), we might have to keep traversing multiple pages to find all commits that we're looking for (say we have a month old PR that got merged).

The current API is also interesting in that, I'm speculating here, but it lists PRs _related_ to the commit. As such I'm thinking _maybe_ it would also list a PR that has been merged into a PR that then is merged into `main`. This could be useful.

But alas, that benefit is theoretical and ultimately, in most cases, not beneficial. The pulls endpoint will likely result in less requests so maybe we go with that.

Comment (deansheather) [reply scripts/release/check_commit_metadata.sh:63]: > do we run dry-run for every merged commit

We do not

What if you instead listed all PRs that had the breaking label and then checked if the merge commit is from any of those PRs?

Comment (mafredri) [reply scripts/release/check_commit_metadata.sh]: @deansheather this is now implemented in [`c871cb7` (#5366)](https://github.com/coder/coder/pull/5366/commits/c871cb7744212652fb0c16c3ca5b7b46546f02fe).

I opted to _not_ filter the search results based on a single tag because I wanted to create a safety-check that ensures we fetched PR data for each commit.

**Review by deansheather — APPROVED** [review]
